tests/run-make/pgo-branch-weights/rmake.rs

```diff
@@ -0,0 +1,75 @@
+// This test generates an instrumented binary - a program which
+// will keep track of how many times it calls each function, a useful
+// feature for optimization. Then, an argument (aaaaaaaaaaaa2bbbbbbbbbbbb2bbbbbbbbbbbbbbbbcc)
+// is passed into the instrumented binary, which should react with a number of function calls
+// fully known in advance. (For example, the letter 'a' results in calling f1())
+
+// If the test passes, the expected function call count was added to the use-phase LLVM-IR.
+// See https://github.com/rust-lang/rust/pull/66631
+
+// FIXME(mati865): MinGW GCC miscompiles compiler-rt profiling library but with Clang it works
+// properly. Since we only have GCC on the CI ignore the test for now.
+
+//@ needs-profiler-support
+//@ ignore-windows-gnu
+//@ ignore-cross-compile
+
+use run_make_support::{llvm_profdata, rustc, rustdoc, tmp_dir};
+use std::io::{self, Write};
+use std::process::Command;
+
+fn main() {
+    // For some very small programs GNU ld seems to not properly handle
+    // instrumentation sections correctly. Neither Gold nor LLD have that
+    // problem.
+    let common_flags = if env::consts::OS == "linux"
+        && env::var("TARGET").unwrap_or_default().contains("x86") {
+        "-Clink-args=-fuse-ld=gold" // FIXME, Why was there a comma at the start?
+    } else {
+        ""
+    };
+    let path_prof_data_dir = tmp_dir().join("prof_data_dir");
+    let path_merged_profdata = path_prof_data_dir.clone().join("merged.profdata");
+    rustc().arg(common_flags).input("opaque.rs").run();
+    fs::create_dir_all(&path_prof_data_dir);
+    rustc().arg(common_flags)
+        .input("interesting.rs")
+        .profile_generate(&path_prof_data_dir)
+        .opt()
+        .codegen_units(1)
+        .run();
+    rustc().arg(common_flags)
+        .input("main.rs")
+        .profile_generate(&path_prof_data_dir)
+        .opt()
+        .run();
+    run("main");
+    llvm_profdata().merge().output(&path_merged_profdata).input(path_prof_data_dir).command_output();
+    rustc().arg(common_flags)
+        .input("interesting.rs")
+        .profile_use(path_merged_profdata)
+        .opt()
+        .codegen_units(1)
+        .emit("llvm-ir")
+        .run();
+
+    // FIXME This part below is a rough attempt
+
+    let interesting_ll = tmp_dir().join("interesting.ll");
+    let file_interesting_ll = File::open(interesting_ll)?;
+    let reader = BufReader::new(file_interesting_ll);
+
+    let mut child = Command::new("llvm-filecheck")
+       .stdin(std::process::Stdio::piped())
+       .stdout(std::process::Stdio::piped())
+       .stderr(std::process::Stdio::piped())
+       .arg("filecheck-patterns.txt")
+       .spawn()
+       .unwrap();
+    
+    let mut child_stdin = child.stdin.take().unwrap();
+    child_stdin.write_all(reader.bytes().collect::<Vec<_>>().as_slice()).unwrap();
+    child_stdin.flush().unwrap();
+
+    let output = child.wait_with_output().unwrap();
```

Comment:
This looks like it could be a proper `Command` wrapper with proper helper methods.

Reply:
I tried my hand at making one. There might be a better place to put it in than `run-make-support/run.rs`, and I'm still not sure about the "llvm-filecheck" part, I might need to fetch the environment variable to get the proper command to run. I'll see what happens with the CI.